source/opt/inline_opaque_pass.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inline_opaque_pass.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+  const uint32_t kTypePointerTypeIdInIdx = 1;
+
+} // anonymous namespace
+
+bool InlineOpaquePass::IsOpaqueType(uint32_t typeId) {
+  const ir::Instruction* typeInst = def_use_mgr_->GetDef(typeId);
+  switch (typeInst->opcode()) {
+    case SpvOpTypeSampler:
+    case SpvOpTypeImage:
+    case SpvOpTypeSampledImage:
+      return true;
+    case SpvOpTypePointer:
+      return IsOpaqueType(typeInst->GetSingleWordInOperand(
+          kTypePointerTypeIdInIdx));
+    default:
+      break;
+  }
+  if (typeInst->opcode() != SpvOpTypeStruct)
```

Comment:
What about Array and RuntimeArray?  

Ok, RuntimeArray would be impossible in the shading langauges I know.  But *perhaps* Array.
Recommend a TODO here, rather than any code change.    Because if you did add Array, then I'd ask for a test. :-)